webapp/src/Service/BalloonService.php

```diff
@@ -144,41 +136,71 @@ public function collectBalloonTable(Contest $contest, bool $todo = false): array
             $balloon = $balloonsData[0];
             $done = $balloon->getDone();
 
-            if ($todo && $done) {
-                continue;
+            // Ensure a summary-row exists for this sortorder and take a reference to these summaries. References are needed to ensure array reuse.
+            // Summaries are used to determine whether a balloon has been handed out so they need to be separated between sortorders.
+            $balloonSummaryPerTeam[$balloonsData['sortorder']] ??= [];
+            $relevantBalloonSummaries = &$balloonSummaryPerTeam[$balloonsData['sortorder']];
+
+            // Commonly no balloons are handed out post freeze.
+            // Underperforming teams' moral can be boosted by handing out balloons post-freeze.
+            // Handing out balloons for problems that have not been solved pre-freeze poses a potential information leak, so these are always excluded.
+            // So to decide whether to skip showing a balloon:
+            //   1. Check whether the scoreboard has been frozen.
+            //   2. Check whether the team has exceeded minimum number of balloons.
+            //   3. Check whether the problem been solved pre-freeze.
+            $stime = $balloonsData['submittime'];
+            if ($ignorePreFreezeSolves === false && isset($freezetime) && $stime >= $freezetime) {
+                if (count($relevantBalloonSummaries[$balloonsData['teamid']] ?? []) >= $minimumNumberOfBalloons) {
+                    continue;
+                }
+
+                // Check if problem has been solved before the freeze by someone in the same sortorder to prevent information leak.
+                // The DOMjudge team (that commonly runs jury submissions) has so must be ignored.
```

Comment:
This sentence looks like it was rewritten partly at some point. What should it say, as I don't see how it handles the DOMjudge team as an edge case?

Reply:
Yeah this reads a bit weird. I think I wrote this in one go. Will rephrase.